### DISCUSSION TITLE
site-to-site VPN; iptables MASQUERADE issue

### DISCUSSION THREAD

**Discussion by Smuger** [strongswan/strongswan#2587] [Q&A]
Hi strongSwan Guys,

![fake_public_site2site drawio](https://github.com/user-attachments/assets/c5e3cf1a-d6c9-4266-97ad-913550c18c4e)

# What's my problem?
Ping **from Cloud** to a random box in on-prem works ✅

Ping **from random box on-prem** to cloud doesn't work ❌

# Investigation

#### Network interaces:
- ens3 for lan
- ipsec0 for the tunnel (This is a **XFRM** Interface)

#### Versions:
strongSwan: 6.0.0
OS: Ubuntu 24.04

### iptables-save
```
# Generated by iptables-save v1.8.10 (nf_tables) on Sat Dec 28 20:44:05 2024
*filter
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [18:1440]
:OUTPUT ACCEPT [0:0]
-A FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
COMMIT
# Completed on Sat Dec 28 20:44:05 2024
# Generated by iptables-save v1.8.10 (nf_tables) on Sat Dec 28 20:44:05 2024
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [772:55276]
-A POSTROUTING -s 172.31.0.0/16 -o ens3 -j MASQUERADE
COMMIT
# Completed on Sat Dec 28 20:44:05 2024
```

### ip a
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute
       valid_lft forever preferred_lft forever
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8950 qdisc fq_codel state UP group default qlen 1000
    link/ether fa:16:3e:75:99:98 brd ff:ff:ff:ff:ff:ff
    altname enp0s3
    inet 10.0.1.103/24 metric 100 brd 10.0.1.255 scope global dynamic ens3
       valid_lft 8315sec preferred_lft 8315sec
    inet6 fe80::f816:3eff:fe75:9998/64 scope link
       valid_lft forever preferred_lft forever
4: ipsec0@NONE: <NOARP,UP,LOWER_UP> mtu 1436 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none
    inet 10.0.1.103 peer 169.254.52.182/30 scope global ipsec0
       valid_lft forever preferred_lft forever
    inet6 fe80::a56b:59cf:547e:75b3/64 scope link stable-privacy
       valid_lft forever preferred_lft forever
```

### conntrack -L
```
icmp     1 29 src=10.0.1.103 dst=172.31.21.160 type=8 code=0 id=24 src=172.31.21.160 dst=10.0.1.103 type=0 code=0 id=24 mark=0 use=1
```

### sudo sysctl -p
```
net.ipv4.ip_forward = 1
```

#### Steps
1. Pinged from on-prem to cloud
2. ICMP reached box in the cloud and reply was send back
3. On strongSwan **ipsec** interface ICMP reply arrived with **Source Address**: 172.31.21.160
4. On strongSwan **ens3** interface ICMP was send away still with  **Source Address**: 172.31.21.160

MASQUERADE should have changed **Source Address** for packages leaving ens3 :(

I bet **Reverse Path Filtering** doesn't like a random 172.31.0.0/16 package reaching my random box. 
- This isn't just a problem with ICMP. 
- Reply from everything else never reaches random box on on-prem. 
- UFW is turned off. Security groups are fully opened. 
- Forwarding is turned on.

Many thanks,
Kryz

**Comment by tobiasbrunner** [comment]
> On strongSwan **ipsec** interface ICMP reply arrived with **Source Address**: 172.31.21.160

What's that "ipsec" interface? An XFRM interface? Something else?

> On strongSwan **ens3** interface ICMP was send away still with  **Source Address**: 172.31.21.160

There is maybe a conntrack entry due to the request from `10.0.1.243` to `172.31.21.160`. So this

> MASQUERADE should have changed **Source Address** for packages leaving ens3 :(

might be the result.

> I bet **Reverse Path Filtering** doesn't like a random 172.31.0.0/16 package reaching my random box.

What do you mean? You were able to route/send a packet to it, so why would the response be blocked?

**Comment by Smuger** [comment]
Hi @tobiasbrunner 
Thank you so much for your answer. I managed to fix this today. I've been using VMs from a third party and they were dropping my traffic tunneled via strongSwan. 

After they turn some services off the tunnel is working great now.

Thank you again
